### PR TITLE
JDK-8318476: Add resource consumption note to BigInteger and BigDecimal

### PR DESCRIPTION
**Pull request by jddarcy**
Add informative notes to BigInteger and BigDecimal about possible running times, etc.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8318476](https://bugs.openjdk.org/browse/JDK-8318476): Add resource consumption note to BigInteger and BigDecimal (**Enhancement** - P4)

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to [c959f3b5](https://git.openjdk.org/jdk/pull/16298/files/c959f3b5282495d90dec00df65aff5bef9fa2f77)
 * [Brian Burkhalter](https://openjdk.org/census#bpb) (@bplb - **Reviewer**) ⚠️ Review applies to [0a48c48f](https://git.openjdk.org/jdk/pull/16298/files/0a48c48f5c07c6fee5e78337f0d66a4b000ccb1f)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16298/head:pull/16298` \
`$ git checkout pull/16298`

Update a local copy of the PR: \
`$ git checkout pull/16298` \
`$ git pull https://git.openjdk.org/jdk.git pull/16298/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16298`

View PR using the GUI difftool: \
`$ git pr show -t 16298`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16298.diff">https://git.openjdk.org/jdk/pull/16298.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16298#issuecomment-1773577195)

### CODE REVIEW

**Review by AlanBateman — APPROVED** [review]
The wording looks okay to me and I expect it will appear as an API note in BigInteger.

However, for BigDecimal I suspect the h2 headings (for the next heading and the existing heading for IEEE 754 Decimal Arithmetic) means they won't look like they are part of the API note, instead they will look like new sections after the API note. Do you see what I mean in the generated javadoc?

File: src/java.base/share/classes/java/math/BigInteger.java [hunk 1]
```diff
@@ -118,6 +118,38 @@
  * the full supported positive range of {@code BigInteger}.
  * The range must be at least 1 to 2<sup>500000000</sup>.
  *
+ * @apiNote
+ * <a id=algorithmicComplexity>As {@code BigInteger} values are
+ * arbitrary precision integers, the algorithmic complexity of the
+ * methods of this class varies and may be superlinear in the size of
+ * the input. For example, a method like {@link intValue()} would be
+ * expected to run in <i>O</i>(1), that is constant time, since with
+ * the current internal presentation only a fixed-size component of
```
Comment (rgiulietti) [inline src/java.base/share/classes/java/math/BigInteger.java:127]:
```suggestion
 * the current internal representation only a fixed-size component of
```

File: src/java.base/share/classes/java/math/BigInteger.java [hunk 2]
```diff
@@ -118,6 +118,38 @@
...
+ * <p>A particular implementation of {@link multiply(BigInteger)
+ * multiply} is free to switch between different algorithms for
+ * different inputs, such as to improve actual running time to produce
+ * the product by using simpler algorithms for smaller inputs even if
+ * the simpler algorithm has a larger asymptotic complexity.
+ *
+ * <p>Users of {@code BigInteger} concerned with bounding the running
+ * time of operations can screen out {@code BigInteger} values above a
+ * chosen magnitude.
+ *
```
Comment (rgiulietti) [inline src/java.base/share/classes/java/math/BigInteger.java:152]: This explains the execution time characteristics but not the memory consumption aspects of the class.
I think that a couple of sentences about them would make this note more complete.

File: src/java.base/share/classes/java/math/BigDecimal.java [hunk 3]
```diff
@@ -292,6 +292,22 @@
...
+ * same {@linkplain BigInteger##algorithmicComplexity complexity
+ * constraints} as {@code BigInteger} multiply of the unscaled
+ * values. In contrast, a {@code BigDecimal} value with a compact
+ * representation like {@code new BigDecimal(1E-1000000000)} has a
+ * {@link toPlainString} result with over one billion characters.
+ *
+ * <p>Users of {@code BigDecimal} concerned with bounding the running
+ * time of operations can screen out {@code BigDecimal} values with
+ * unscaled values or scales above a chosen magnitude.
+ *
```
Comment (rgiulietti) [inline src/java.base/share/classes/java/math/BigDecimal.java:310]: A paragraph about memory consumption is perhaps worthwhile as well.

File: src/java.base/share/classes/java/math/BigInteger.java (hunk 2 above)
Comment (jddarcy) [reply src/java.base/share/classes/java/math/BigInteger.java:152]: A reasonable suggestion; I was considering mentioning memory usage in the initial version. I'll craft something brief for both classes on memory usage.

File: src/java.base/share/classes/java/math/BigInteger.java [hunk 4]
```diff
@@ -124,7 +124,7 @@
  * methods of this class varies and may be superlinear in the size of
  * the input. For example, a method like {@link intValue()} would be
  * expected to run in <i>O</i>(1), that is constant time, since with
- * the current internal presentation only a fixed-size component of
+ * the current internal reprentation only a fixed-size component of
```
Comment (rgiulietti) [inline src/java.base/share/classes/java/math/BigInteger.java:127]:
```suggestion
 * the current internal representation only a fixed-size component of
```

Comment (rgiulietti) [reply src/java.base/share/classes/java/math/BigInteger.java:127]: LGTM

**Comment by bplb** [comment]
I suggest taking a look at [JDK-8225458](https://bugs.openjdk.org/browse/JDK-8225458) while in this area.

File: src/java.base/share/classes/java/math/BigDecimal.java [hunk 5]
```diff
@@ -292,6 +292,27 @@
...
  * {@code BigDecimal} arithmetic.
  *
+ * <h2><a id=algorithmicComplexity>Algorithmic Complexity</a></h2>
+ *
+ * Operations on {@code BigDecimal} values have a range of algorithmic
+ * complexities; in general, those complexities are a function of both
+ * the size of the unscaled value as well as the size of the
+ * scale. For example, an {@linkplain BigDecimal#multiply(BigDecimal)
+ * exact multiply} of two {@code BigDecimal} values is subject to the
+ * same {@linkplain BigInteger##algorithmicComplexity complexity
```
Comment (bplb) [inline src/java.base/share/classes/java/math/BigDecimal.java:302]: Does this link need two `##` or only one?

File: src/java.base/share/classes/java/math/BigInteger.java [hunk 6]
```diff
@@ -118,6 +118,42 @@
...
+ * the {@code BigInteger} needs to be accessed to perform the
+ * conversion to {@code int}. In contrast, a method like {@link not()}
+ * would be expected to run in <i>O</i>(<i>n</i>) time where <i>n</i>
+ * is size of the {@code BigInteger} in bits, that is, to run in time
+ * proportional to the size of the input. For multiplying two {@code
+ * BigInteger} values of size <i>n</i>, a naive multiply algorithm
+ * would run in time <i>O</i>(<i>n<sup>2</sup></i>) and theoretical
+ * results indicate a multiply algorithm for numbers using this
+ * category of representation must run in <em>at least</em>
+ * <i>O</i>(<i>n</i>&nbsp;log&nbsp;<i>n</i>). Common multiply
```
Comment (bplb) [inline src/java.base/share/classes/java/math/BigInteger.java:137]: "Common _multiply_" or "Common _multiplication_"?

File: src/java.base/share/classes/java/math/BigDecimal.java (hunk 5 above)
Comment (jddarcy) [reply src/java.base/share/classes/java/math/BigDecimal.java:302]: The "##" is used for HTML id's declared in the javadoc.

File: src/java.base/share/classes/java/math/BigInteger.java (hunk 6 above)
Comment (jddarcy) [reply src/java.base/share/classes/java/math/BigInteger.java:137]: Doing some grammar research ...

In English, using a noun as an adjective is called an "attributive noun," so the "multiplication" in "multiplication algorithm" would play that role.

Strictly speaking to "adjective-ifly" a verb would be to use a participle, one way to do that anyway -- "multiplying algorithm."

In any case, "multiply algorithm" reads okay to me, but I don't mind changing it to "multiplication algorithm."

File: src/java.base/share/classes/java/math/BigDecimal.java (hunk 5 above)
Comment (bplb) [reply src/java.base/share/classes/java/math/BigDecimal.java:302]: Okay, thanks.

File: src/java.base/share/classes/java/math/BigInteger.java (hunk 6 above)
Comment (bplb) [reply src/java.base/share/classes/java/math/BigInteger.java]: Then it's fine as it is.

**Review by bplb — APPROVED** [review]

**Pull request closed without merging**